### PR TITLE
Fix the msecs part printing of the time strings in the logs [API-560]

### LINKED ISSUE #892: Connection log timestamp contains negative msecs
@ihsandemir  Hi - I'll test that tomorrow. Also - in the last read / last write time area - I got an odd log from the library with a negative millisecond timestamp - see "lastReadTime=2021-06-23 15:35:58.-67" in the log below.

```
2021/06/23-15:35:58.768856 +0000: (90): INFO: Discovery: Library - <hz.client_1><soaktest> ClientConnection{alive=0, connectionId=5, remoteEndpoint=Address[10.42.4.229:5701], lastReadTime=2021-06-23 15:35:58.-67, closedTime=2021-06-23 15:35:58.000, connected server version=4.0.3} closed. Reason: Socket read error. asio.misc:2 for ClientConnection{alive=1, connectionId=5, remoteEndpoint=Address[10.42.4.229:5701], lastReadTime=2021-06-23 15:35:58.-67, closedTime=never, connected server version=4.0.3}
.0.3}
```

_Originally posted by @andysCaplin in https://github.com/hazelcast/hazelcast-cpp-client/issues/883#issuecomment-866994276_

This seems to happen if the timestamp is before the current time in the system.

### PR DESCRIPTION
**Pull request by ihsandemir**
Fixed the time printing for `Connection` logs so that the msecs part is correctly written.

fixes #892

### CODE REVIEW

File: hazelcast/test/src/HazelcastTests2.cpp [hunk 1]
```diff
@@ -191,6 +191,14 @@ namespace hazelcast {
                 int64_t actualDuration = end - start;
                 ASSERT_GE(actualDuration, parkDurationNanos);
             }
+
+            TEST_F (ClientUtilTest, print_older_stead_clock_time) {
```
Comment (yuce) [inline hazelcast/test/src/HazelcastTests2.cpp:195]: Should it be: `steady_clock_time` ?

**Review by yuce — DISMISSED** [review]
Looks good, except a typo.

File: hazelcast/test/src/HazelcastTests2.cpp (hunk 1 above)
Comment (ihsandemir) [reply hazelcast/test/src/HazelcastTests2.cpp]: fixed

File: hazelcast/src/hazelcast/util/util.cpp [hunk 2]
```diff
@@ -445,7 +452,7 @@ namespace hazelcast {
             char time_buffer[80];
             std::strftime(time_buffer, sizeof(time_buffer), "%Y-%m-%d %H:%M:%S", &localBrokenTime);
             oss << time_buffer;
-            oss << '.' << std::setfill('0') << std::setw(3) << duration_cast<milliseconds>(systemDuration).count() % 1000;
+            oss << '.' << msecs;
```
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/src/hazelcast/util/util.cpp]: Fixed.

File: hazelcast/src/hazelcast/util/util.cpp [hunk 3]
```diff
@@ -435,7 +435,14 @@ namespace hazelcast {
...
             auto systemDuration = duration_cast<system_clock::duration>(t - steady_clock::now());
-            auto brokenTime = system_clock::to_time_t(system_clock::now() + systemDuration);
+            auto msecs = duration_cast<milliseconds>(systemDuration).count() % 1000;
+            auto system_time = system_clock::now() + systemDuration;
+            if (msecs < 0) {
+                msecs = 1000 + msecs;
+                system_time -= std::chrono::seconds(1);
+            }
+
+            auto brokenTime = system_clock::to_time_t(system_time);
```
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/src/hazelcast/util/util.cpp]: fixed.

File: hazelcast/test/src/HazelcastTests2.cpp [hunk 4]
```diff
@@ -172,7 +172,7 @@ namespace hazelcast {
                 std::string timeString = hazelcast::util::StringUtil::time_to_string(
                         std::chrono::steady_clock::now());
                 //expected format is "%Y-%m-%d %H:%M:%S.%f" it will be something like 2018-03-20 15:36:07.280
-                ASSERT_EQ((size_t) 23, timeString.length());
+                ASSERT_LE((size_t) 21, timeString.length());
```
Comment (yemreinci) [inline hazelcast/test/src/HazelcastTests2.cpp:175]:
```suggestion
                ASSERT_EQ(static_cast<std::size_t>(23), timeString.length());
```
setw will now ensure that the length is always 23.

**Review by yemreinci — APPROVED** [review]
Approved with one suggestion.

**Comment by yemreinci** [comment]
verify-linux

**Review by yemreinci — APPROVED** [review]

**Pull request merged**